### ISSUE TITLE
Lag  when using 20px brush on a 512x512 canvas caused many calls inside draw_fill_gap 

### ISSUE THREAD

**Issue by Nagamine-j** [Orama-Interactive/Pixelorama#501]
**Pixelorama version:**
5b74a4292d874bc69c76adf8e3c7ddbbb8b4dd2e

**OS/device including version:**
Windows 10 / laptop ryzen 5 amd vega 8

**Issue description:**

Lag when using 20px brush on a 512x512 canvas, It's cause because the 50 000 calls to can pixel get wich are caused by the draw_fill_gap function
![image](https://user-images.githubusercontent.com/60224203/123057243-51db1200-d3cd-11eb-826f-bcf6289b09e0.png)
https://github.com/Orama-Interactive/Pixelorama/blob/5b74a4292d874bc69c76adf8e3c7ddbbb8b4dd2e/src/Tools/Draw.gd#L202-L221

looking at the algorithm on  https://godotengine.org/qa/35276/tile-based-line-drawing-algorithm-efficiency
I think the draw_tool(Vector2(x, y))  function call should be outside the while

**Steps to reproduce:**
1. Create a 512 x512 project
2. select 20px brush
3. draw

**Comment by Nagamine-j** [comment]
Ok i checked again and the draw_tool function call goes inside the while, but it calls draw_tool incrementing the diferential pixel by pixel and every draw_pixel calls set_pixel 400 times (which is understandable because the brush is 20*20) however many of this function calls are unncesary because most of the pixels are already set in the previews draw_pixel call.
I think it should be more optimized approach

**Comment by Docmine17** [comment]
I have the same problem here, I use Ubuntu 21.04.

**Comment by Variable-ind** [comment]
Hy :wave:, i tried to optimize it a bit [here's](https://nightly.link/Orama-Interactive/Pixelorama/actions/runs/1918042931) the nightly link to it. Does this solve the problem?

**Comment by MatteoPiovanelli-Laser** [comment]
Hello.
I did some tests on this. The `_cache_size` as it's being computed seems to be helpful when the brush type is "Pixel brush". For other types its usefulness in making things faster varies. I changed the `_set_pixel(position: Vector2, ignore_mirroring := false)` method as follows:

```
func _set_pixel(position: Vector2, ignore_mirroring := false) -> void:
	if position in _draw_cache and _for_frame == Global.current_project.current_frame:
		return
	_update_brush_cache(position)

	var project: Project = Global.current_project
	if project.tile_mode and project.get_tile_mode_rect().has_point(position):
		position = position.posmodv(project.size)

	if !project.can_pixel_get_drawn(position):
		return

	var images := _get_selected_draw_images()
	var i := int(position.x + position.y * project.size.x)
	if _mask.size() >= i + 1:
		if _mask[i] < Tools.pen_pressure:
			_mask[i] = Tools.pen_pressure
			for image in images:
				_drawer.set_pixel(image, position, tool_slot.color, ignore_mirroring)
	else:
		for image in images:
			_drawer.set_pixel(image, position, tool_slot.color, ignore_mirroring)


func _update_brush_cache(position: Vector2) -> void:
	if _draw_cache.size() > _cache_limit or _for_frame != Global.current_project.current_frame:
		_draw_cache = []
		_for_frame = Global.current_project.current_frame
	_draw_cache.append(position)  # Store the position of pixel
```
The point of separating out `_update_brush_cache(position: Vector2)` is better "see" cache misses on the profiler.

On a 256x256 image, with a "Filled circle brush" with size 20px, here's what I measured:
![image](https://user-images.githubusercontent.com/19283945/170140747-d660f60e-f16a-4f79-8b31-7cfa5522b642.png)
As you can see from the two lines I marked, for most pixels the cache is being missed, which makes sense because it's much smaller than the size of the tool. The size of that should probably depend on the type of tool being used.

On the other hand, there are a few things that stand out to me from those profile traces.

- `_get_selected_draw_images()` is being called for every pixel. Reasonably, I would expect the selected layer and frame to not change during a single "draw" operation, or at least for the duration of a `draw_tool`. Are there cases where this isn't guaranteed? 
- `if _mask.size() >= i + 1:` similar considerations could probably be made for that entire branch of that function. Looks to me that `_maks.size()` is either `0` or the size in pixels of the image / project. At least the first condition could be detected early to invoke a more optimized `_set_pixel()` that doesn't need to test for that.
- `if project.tile_mode` same considerations as above.

Basically, the last two points would mean that the condition `tile_mode == 0 and _mask.size() == 0`, and each separate of those conditions, could have shortcuts to make them faster, while the first point would let us cut a bunch of operations off this process entirely.

**Comment by MatteoPiovanelli-Laser** [comment]
https://github.com/Orama-Interactive/Pixelorama/pull/705#pullrequestreview-1002774685

copying link to that comment here as a reminder for myself (or whoever else may want to pick this up, I'm not jealous).
I have my hands full with other stuff for at least a week or so before I can give any attention to this.

**Comment by MatteoPiovanelli-Laser** [comment]
![image](https://user-images.githubusercontent.com/19283945/179099119-0368b238-2184-45e9-b562-d13421438a67.png)



I can do a further pass on this. I think I'll start by removing a bunch of repeated calls whose results should be memorized rather than recomputed on every step we are taking while drawing the pencil "segments".
First example of this I think would be the results of `_prepare_circle_tool`

**Comment by MatteoPiovanelli-Laser** [comment]
I've been playing with this recently and I don't think this is really an issue any more.
Godot 3.5 definitely fixed something major behind the scenes.

https://user-images.githubusercontent.com/19283945/188727151-bd32d5c3-e319-4b14-9dd7-b406be979bb1.mp4

(image is 512x512px. Brush size is 25 px)
